// @iplab/ngx-color-picker/esm2015/lib/components/parts/alpha/alpha.component.js
import { Component, Input, EventEmitter, Output, ChangeDetectionStrategy, Inject, ElementRef, ViewChild, HostListener, Renderer2 } from '@angular/core';
import { DOCUMENT } from '@angular/common';
import { Color } from './../../../helpers/color.class';
import { BaseComponent } from './../base.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

const _c0 = ["pointer"];
const _c1 = function (a0) { return { "background": a0 }; };
export class AlphaComponent extends BaseComponent {
    constructor(renderer, document, elementRef) {
        super(document, elementRef, renderer);
        this.colorChange = new EventEmitter(false);
        this.isVertical = false;
    }
    onClick(event) {
        this.onEventChange(event);
    }
    set vertical(value) {
        this.isVertical = true;
    }
    /**
     * color can be changed through inputs
     * and then we need to move pointer
     */
    ngOnChanges(changes) {
        if (changes.color && changes.color.previousValue !== changes.color.currentValue) {
            const hsva = this.color.getHsva();
            this.changePointerPosition(hsva.alpha);
        }
    }
    movePointer({ x, y, height, width }) {
        const alpha = this.isVertical ? y / height : x / width;
        this.changePointerPosition(alpha);
        const hsva = this.color.getHsva();
        const newColor = new Color().setHsva(hsva.hue, hsva.saturation, hsva.value, alpha);
        this.colorChange.emit(newColor);
    }
    /**
     * hue value is in range from 0 to 360°
     */
    changePointerPosition(alpha) {
        const x = alpha * 100;
        const orientation = this.isVertical ? 'top' : 'left';
        this.renderer.setStyle(this.pointer.nativeElement, orientation, `${x}%`);
    }
    get gradient() {
        const rgba = this.color.getRgba();
        const orientation = this.isVertical ? 'bottom' : 'right';
        return `linear-gradient(to ${orientation}, rgba(${rgba.red}, ${rgba.green}, ${rgba.blue}, 0) 0%, rgb(${rgba.red}, ${rgba.green}, ${rgba.blue}) 100%)`;
    }
}
AlphaComponent.ɵfac = function AlphaComponent_Factory(t) { return new (t || AlphaComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(DOCUMENT), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
AlphaComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AlphaComponent, selectors: [["alpha-component"]], viewQuery: function AlphaComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c0, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.pointer = _t.first);
    } }, hostBindings: function AlphaComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("mousedown", function AlphaComponent_mousedown_HostBindingHandler($event) { return ctx.onClick($event); })("touchstart", function AlphaComponent_touchstart_HostBindingHandler($event) { return ctx.onClick($event); });
    } }, inputs: { vertical: "vertical", color: "color" }, outputs: { colorChange: "colorChange" }, features: [ɵngcc0.ɵɵInheritDefinitionFeature, ɵngcc0.ɵɵNgOnChangesFeature], decls: 3, vars: 3, consts: [[1, "pointer"], ["pointer", ""], [1, "gradient-color", 3, "ngStyle"]], template: function AlphaComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelement(0, "div", 0, 1);
        ɵngcc0.ɵɵelement(2, "div", 2);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction1(1, _c1, ctx.gradient));
    } }, directives: [ɵngcc1.NgStyle], styles: ["[_nghost-%COMP%], [_nghost-%COMP%]     *{padding:0;margin:0;box-sizing:border-box}", "[_nghost-%COMP%]{display:block;background-image:url(\"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAMUlEQVQ4T2NkYGAQYcAP3uCTZhw1gGGYhAGBZIA/nYDCgBDAm9BGDWAAJyRCgLaBCAAgXwixzAS0pgAAAABJRU5ErkJggg==\");background-position:0;height:12px;border-radius:2px;position:relative}[vertical][_nghost-%COMP%]{width:12px;height:100px;background-position:center 0}.gradient-color[_ngcontent-%COMP%]{position:absolute;left:0;right:0;top:0;height:100%;z-index:1}.pointer[_ngcontent-%COMP%]{background:#fff;height:14px;width:14px;top:-1px;left:0;position:absolute;border-radius:50%;cursor:pointer;margin:0 0 0 -7px;z-index:2}[vertical][_nghost-%COMP%]   .pointer[_ngcontent-%COMP%]{left:-1px;margin:-7px 0 0}"], changeDetection: 0 });
AlphaComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: ElementRef }
];
AlphaComponent.propDecorators = {
    color: [{ type: Input }],
    colorChange: [{ type: Output }],
    pointer: [{ type: ViewChild, args: ['pointer', { static: true },] }],
    onClick: [{ type: HostListener, args: ['mousedown', ['$event'],] }, { type: HostListener, args: ['touchstart', ['$event'],] }],
    vertical: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AlphaComponent, [{
        type: Component,
        args: [{
                selector: `alpha-component`,
                template: "<div #pointer class=\"pointer\"></div>\r\n<div class=\"gradient-color\" [ngStyle]=\"{ 'background': gradient }\"></div>",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [":host,:host ::ng-deep *{padding:0;margin:0;box-sizing:border-box}", ":host{display:block;background-image:url(\"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAMUlEQVQ4T2NkYGAQYcAP3uCTZhw1gGGYhAGBZIA/nYDCgBDAm9BGDWAAJyRCgLaBCAAgXwixzAS0pgAAAABJRU5ErkJggg==\");background-position:0;height:12px;border-radius:2px;position:relative}:host([vertical]){width:12px;height:100px;background-position:center 0}.gradient-color{position:absolute;left:0;right:0;top:0;height:100%;z-index:1}.pointer{background:#fff;height:14px;width:14px;top:-1px;left:0;position:absolute;border-radius:50%;cursor:pointer;margin:0 0 0 -7px;z-index:2}:host([vertical]) .pointer{left:-1px;margin:-7px 0 0}"]
            }]
    }], function () { return [{ type: ɵngcc0.Renderer2 }, { type: undefined, decorators: [{
                type: Inject,
                args: [DOCUMENT]
            }] }, { type: ɵngcc0.ElementRef }]; }, { colorChange: [{
            type: Output
        }], onClick: [{
            type: HostListener,
            args: ['mousedown', ['$event']]
        }, {
            type: HostListener,
            args: ['touchstart', ['$event']]
        }], vertical: [{
            type: Input
        }], color: [{
            type: Input
        }], pointer: [{
            type: ViewChild,
            args: ['pointer', { static: true }]
        }] }); })();
