// @iplab/ngx-color-picker/esm2015/lib/components/parts/color-presets/color-presets.component.js
import { Component, Input, EventEmitter, Output, ChangeDetectionStrategy } from '@angular/core';
import { Color } from './../../../helpers/color.class';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '../color-preset-sublist/color-preset-sublist.component';
import * as ɵngcc3 from '../color-preset/color-preset.component';
import * as ɵngcc4 from '../../../pipes/chunks.pipe';

const _c0 = function (a0, a1) { return { "first": a0, "last": a1 }; };
function ColorPresetsComponent_div_0_ng_template_1_color_preset_sublist_0_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "color-preset-sublist", 5);
    ɵngcc0.ɵɵlistener("selectionChange", function ColorPresetsComponent_div_0_ng_template_1_color_preset_sublist_0_Template_color_preset_sublist_selectionChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r11 = ɵngcc0.ɵɵnextContext(3); return ctx_r11.onSelectionChange($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r13 = ɵngcc0.ɵɵnextContext();
    const preset_r5 = ctx_r13.$implicit;
    const first_r6 = ctx_r13.first;
    const last_r7 = ctx_r13.last;
    const ctx_r8 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("list", preset_r5)("direction", ctx_r8.direction)("activeColor", ctx_r8.color)("ngClass", ɵngcc0.ɵɵpureFunction2(4, _c0, first_r6, last_r7));
} }
function ColorPresetsComponent_div_0_ng_template_1_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    const _r15 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "color-preset", 6);
    ɵngcc0.ɵɵlistener("selectionChange", function ColorPresetsComponent_div_0_ng_template_1_ng_template_1_Template_color_preset_selectionChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r15); const ctx_r14 = ɵngcc0.ɵɵnextContext(3); return ctx_r14.onSelectionChange($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r16 = ɵngcc0.ɵɵnextContext();
    const first_r6 = ctx_r16.first;
    const last_r7 = ctx_r16.last;
    const preset_r5 = ctx_r16.$implicit;
    const ctx_r10 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(3, _c0, first_r6, last_r7))("color", preset_r5)("activeColor", ctx_r10.color);
} }
function ColorPresetsComponent_div_0_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ColorPresetsComponent_div_0_ng_template_1_color_preset_sublist_0_Template, 1, 7, "color-preset-sublist", 3);
    ɵngcc0.ɵɵtemplate(1, ColorPresetsComponent_div_0_ng_template_1_ng_template_1_Template, 1, 6, "ng-template", null, 4, ɵngcc0.ɵɵtemplateRefExtractor);
} if (rf & 2) {
    const preset_r5 = ctx.$implicit;
    const _r9 = ɵngcc0.ɵɵreference(2);
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.isList(preset_r5))("ngIfElse", _r9);
} }
function ColorPresetsComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 1);
    ɵngcc0.ɵɵtemplate(1, ColorPresetsComponent_div_0_ng_template_1_Template, 3, 2, "ng-template", 2);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const chunk_r1 = ctx.$implicit;
    const first_r2 = ctx.first;
    const last_r3 = ctx.last;
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(2, _c0, first_r2, last_r3));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", chunk_r1);
} }
export class ColorPresetsComponent {
    constructor() {
        this.columns = 8;
        this.hueChange = new EventEmitter(false);
        this.colorChange = new EventEmitter(false);
        this.direction = 'up';
    }
    onSelectionChange(color) {
        const selectedRgbaColor = color.getRgba();
        const selectedHsvaColor = color.getHsva();
        const newColor = new Color().setRgba(selectedRgbaColor.red, selectedRgbaColor.green, selectedRgbaColor.blue, selectedRgbaColor.alpha);
        const hueColor = new Color().setHsva(selectedHsvaColor.hue);
        this.hueChange.emit(hueColor);
        this.colorChange.emit(newColor);
    }
    isList(colorPreset) {
        return Array.isArray(colorPreset);
    }
}
ColorPresetsComponent.ɵfac = function ColorPresetsComponent_Factory(t) { return new (t || ColorPresetsComponent)(); };
ColorPresetsComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ColorPresetsComponent, selectors: [["color-presets-component"]], inputs: { columns: "columns", direction: "direction", colorPresets: "colorPresets", hue: "hue", color: "color" }, outputs: { hueChange: "hueChange", colorChange: "colorChange" }, decls: 2, vars: 4, consts: [["class", "row", 3, "ngClass", 4, "ngFor", "ngForOf"], [1, "row", 3, "ngClass"], ["ngFor", "", 3, "ngForOf"], [3, "list", "direction", "activeColor", "ngClass", "selectionChange", 4, "ngIf", "ngIfElse"], ["colorPreset", ""], [3, "list", "direction", "activeColor", "ngClass", "selectionChange"], [3, "ngClass", "color", "activeColor", "selectionChange"]], template: function ColorPresetsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ColorPresetsComponent_div_0_Template, 2, 5, "div", 0);
        ɵngcc0.ɵɵpipe(1, "chunks");
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngForOf", ɵngcc0.ɵɵpipeBind2(1, 1, ctx.colorPresets, ctx.columns));
    } }, directives: [ɵngcc1.NgForOf, ɵngcc1.NgClass, ɵngcc1.NgIf, ɵngcc2.ColorPresetSublist, ɵngcc3.ColorPresetComponent], pipes: [ɵngcc4.ChunksPipe], styles: ["[_nghost-%COMP%], [_nghost-%COMP%]     *{padding:0;margin:0;box-sizing:border-box}", "[_nghost-%COMP%]{display:block;font-size:0}[_nghost-%COMP%]   .row[_ngcontent-%COMP%]:first-child{padding:0}color-preset-sublist[_ngcontent-%COMP%]:first-child, color-preset[_ngcontent-%COMP%]:first-child{margin:0}"], changeDetection: 0 });
ColorPresetsComponent.propDecorators = {
    columns: [{ type: Input }],
    colorPresets: [{ type: Input }],
    hue: [{ type: Input }],
    hueChange: [{ type: Output }],
    color: [{ type: Input }],
    colorChange: [{ type: Output }],
    direction: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ColorPresetsComponent, [{
        type: Component,
        args: [{
                selector: `color-presets-component`,
                template: "<div class=\"row\" *ngFor=\"let chunk of colorPresets | chunks: columns; let first = first; let last = last;\" [ngClass]=\"{ 'first': first, 'last': last }\">\r\n    <ng-template ngFor let-preset let-first=\"first\" let-last=\"last\" [ngForOf]=\"chunk\">\r\n        <color-preset-sublist \r\n            [list]=\"preset\" \r\n            *ngIf=\"isList(preset); else colorPreset\" \r\n            [direction]=\"direction\"\r\n            [activeColor]=\"color\"\r\n            [ngClass]=\"{ 'first': first, 'last': last }\"\r\n            (selectionChange)=\"onSelectionChange($event)\"></color-preset-sublist>\r\n\r\n        <ng-template #colorPreset>\r\n            <color-preset \r\n                [ngClass]=\"{ 'first': first, 'last': last }\"\r\n                [color]=\"preset\" \r\n                [activeColor]=\"color\" \r\n                (selectionChange)=\"onSelectionChange($event)\"></color-preset>\r\n        </ng-template>\r\n    </ng-template>\r\n</div>",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [":host,:host ::ng-deep *{padding:0;margin:0;box-sizing:border-box}", ":host{display:block;font-size:0}:host .row:first-child{padding:0}color-preset-sublist:first-child,color-preset:first-child{margin:0}"]
            }]
    }], function () { return []; }, { columns: [{
            type: Input
        }], hueChange: [{
            type: Output
        }], colorChange: [{
            type: Output
        }], direction: [{
            type: Input
        }], colorPresets: [{
            type: Input
        }], hue: [{
            type: Input
        }], color: [{
            type: Input
        }] }); })();
