// mat-video/esm5/lib/ui/mat-time-control/mat-time-control.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from "@angular/core";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../pipes/seconds-to-time.pipe';
var MatTimeControlComponent = /** @class */ (function () {
    function MatTimeControlComponent() {
    }
    /** @nocollapse */
    MatTimeControlComponent.ctorParameters = function () { return []; };
    MatTimeControlComponent.propDecorators = {
        video: [{ type: Input }]
    };
MatTimeControlComponent.ɵfac = function MatTimeControlComponent_Factory(t) { return new (t || MatTimeControlComponent)(); };
MatTimeControlComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: MatTimeControlComponent, selectors: [["mat-time-control"]], inputs: { video: "video" }, decls: 4, vars: 6, consts: [[1, "playtime"]], template: function MatTimeControlComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "secondsToTime");
        ɵngcc0.ɵɵpipe(3, "secondsToTime");
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate2("", ɵngcc0.ɵɵpipeBind1(2, 2, ctx.video == null ? null : ctx.video.currentTime), " / ", ɵngcc0.ɵɵpipeBind1(3, 4, ctx.video == null ? null : ctx.video.duration), "");
    } }, pipes: [ɵngcc1.SecondsToTimePipe], styles: [".playtime[_ngcontent-%COMP%]{display:inline;font-size:12px}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatTimeControlComponent, [{
        type: Component,
        args: [{
                selector: "mat-time-control",
                template: "<div class=\"playtime\">{{ video?.currentTime | secondsToTime }} / {{ video?.duration | secondsToTime }}</div>\r\n",
                styles: [".playtime{display:inline;font-size:12px}"]
            }]
    }], function () { return []; }, { video: [{
            type: Input
        }] }); })();
    return MatTimeControlComponent;
}());
export { MatTimeControlComponent };
if (false) {
    /** @type {?} */
    MatTimeControlComponent.prototype.video;
}
