// ng-zorro-antd/esm5/comment/comment.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: comment.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { __read, __spread } from "tslib";
/**
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NzOutletModule } from 'ng-zorro-antd/core/outlet';
import { NzCommentActionComponent, NzCommentActionHostDirective, NzCommentAvatarDirective, NzCommentContentDirective } from './comment-cells';
import { NzCommentComponent } from './comment.component';
/** @type {?} */
import * as ɵngcc0 from '@angular/core';
var NZ_COMMENT_CELLS = [NzCommentAvatarDirective, NzCommentContentDirective, NzCommentActionComponent, NzCommentActionHostDirective];
var NzCommentModule = /** @class */ (function () {
    function NzCommentModule() {
    }
NzCommentModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NzCommentModule });
NzCommentModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NzCommentModule_Factory(t) { return new (t || NzCommentModule)(); }, imports: [[CommonModule, NzOutletModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NzCommentModule, { declarations: function () { return [NzCommentComponent,
        NzCommentAvatarDirective, NzCommentContentDirective, NzCommentActionComponent, NzCommentActionHostDirective]; }, imports: function () { return [CommonModule, NzOutletModule]; }, exports: function () { return [NzCommentComponent,
        NzCommentAvatarDirective, NzCommentContentDirective, NzCommentActionComponent, NzCommentActionHostDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NzCommentModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, NzOutletModule],
                exports: __spread([NzCommentComponent], NZ_COMMENT_CELLS),
                declarations: __spread([NzCommentComponent], NZ_COMMENT_CELLS)
            }]
    }], function () { return []; }, null); })();
    return NzCommentModule;
}());
export { NzCommentModule };
