// ngx-nestable/esm2015/lib/nestable.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NestableComponent } from './nestable.component';
import { NestableDragHandleDirective } from './nestable-drag-handle/nestable-drag-handle.directive';
import { NestableExpandCollapseDirective } from './nestable-expand-collapse-handle/nestable-expand-collapse.directive';
import * as ɵngcc0 from '@angular/core';
export class NestableModule {
}
NestableModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NestableModule });
NestableModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NestableModule_Factory(t) { return new (t || NestableModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NestableModule, { declarations: function () { return [NestableComponent,
        NestableDragHandleDirective,
        NestableExpandCollapseDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [NestableComponent,
        NestableDragHandleDirective,
        NestableExpandCollapseDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NestableModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [
                    NestableComponent,
                    NestableDragHandleDirective,
                    NestableExpandCollapseDirective
                ],
                exports: [
                    NestableComponent,
                    NestableDragHandleDirective,
                    NestableExpandCollapseDirective
                ]
            }]
    }], null, null); })();
