// @iplab/ngx-color-picker/esm2015/lib/components/swatches-picker/swatches-picker.component.js
import { Component, Input, Output, EventEmitter, ChangeDetectionStrategy, ChangeDetectorRef } from '@angular/core';
import { ColorPickerControl } from './../../helpers/control.class';
import { getValueByType } from './../../helpers/helper.functions';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../parts/color-presets/color-presets.component';
import * as ɵngcc2 from '@angular/common';

function SwatchesPickerComponent_color_presets_component_1_Template(rf, ctx) { if (rf & 1) {
    const _r2 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "color-presets-component", 2);
    ɵngcc0.ɵɵlistener("colorChange", function SwatchesPickerComponent_color_presets_component_1_Template_color_presets_component_colorChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r2); const ctx_r1 = ɵngcc0.ɵɵnextContext(); return ctx_r1.childControl.value = $event; });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("columns", 7)("color", ctx_r0.childControl.value)("colorPresets", ctx_r0.childControl.presets);
} }
export class SwatchesPickerComponent {
    constructor(cdr) {
        this.cdr = cdr;
        this.colorChange = new EventEmitter(false);
        this.control = new ColorPickerControl();
        this.childControl = new ColorPickerControl();
        this.subscriptions = [];
        this.mapColors = {
            '#E6315B': [
                '#fc8da7', '#fa7d9a', '#f56484', '#f04a71', '#e82c58', '#e31746', '#de0235',
                '#d60234', '#d10232', '#c70230', '#b8022c', '#ab0229', '#9c0225', '#8f0122',
                '#8c0122', '#82011f', '#78011b', '#690117', '#5c0012', '#4f0010', '#42000c'
            ],
            '#793183': [
                '#ef8dfc', '#eb7dfa', '#e664f5', '#dc4af0', '#d22ce8', '#cb17e3', '#c402de',
                '#c002d9', '#bb02d4', '#b002c7', '#a202b8', '#9702ab', '#8a029c', '#7e018f',
                '#7a018a', '#730182', '#6c0178', '#5e0169', '#54015c', '#49014f', '#3d0142'
            ],
            '#009DE7': [
                '#8dd9fc', '#7dd2fa', '#64c7f5', '#4abbf0', '#2cade8', '#17a2e3', '#0298de',
                '#0295d9', '#0291d4', '#0289c7', '#027eb8', '#0275ab', '#026b9c', '#01628f',
                '#015f8a', '#015982', '#015278', '#014869', '#013f5c', '#01364f', '#012e42'
            ],
            '#00B59C': [
                '#8dfeea', '#7dfbe4', '#63f4db', '#4befd2', '#2de7c6', '#16e2be', '#03deb7',
                '#01ddb6', '#01d4ae', '#01c7a4', '#01b897', '#01aa8b', '#019b80', '#019076',
                '#018c73', '#01836c', '#017763', '#016857', '#005c4e', '#005044', '#004239'
            ],
            '#FFCE00': [
                '#fce68d', '#fae17d', '#f5da64', '#f0cf4a', '#e8c22c', '#e5bc17', '#deb202',
                '#deb100', '#d4aa02', '#c7a002', '#b89302', '#ab8902', '#9c7d02', '#8f7301',
                '#8c7001', '#826801', '#786201', '#695601', '#5c4b00', '#4f4100', '#423700'
            ],
            '#FF4A21': [
                '#fca28d', '#fa947d', '#f57f64', '#f0694a', '#e84f2c', '#e33c17', '#de2a02',
                '#d92a02', '#d42902', '#c72602', '#b82302', '#ab2102', '#9c1e02', '#8f1b01',
                '#8a1a01', '#821901', '#781701', '#691300', '#5c1100', '#4f0e00', '#420c00'
            ],
            '#D6D5D6': [
                '#fff', '#f2f2f2', '#e5e5e5', '#d9d9d9', '#cccccc', '#bfbfbf', '#b3b3b3',
                '#a6a6a6', '#999999', '#8c8c8c', '#808080', '#737373', '#666666', '#595959',
                '#4d4d4d', '#424242', '#363636', '#262626', '#1a1a1a', '#0f0f0f', '#000'
            ]
        };
    }
    ngOnInit() {
        if (this.color) {
            this.childControl.setValueFrom(this.color);
        }
        else {
            this.control.setValueFrom('#E6315B');
        }
        /**
         * set color presets
         * defined by swatches color picker component
         */
        this.control.setColorPresets([
            '#e6315b', '#793183', '#009de7', '#00b59c', '#ffce00', '#ff4a21', '#d6d5d6'
        ]);
        /**
         * initially open first group
         */
        this.childControl.setColorPresets(this.mapColors['#E6315B']);
        this.subscriptions.push(this.childControl.valueChanges.subscribe((value) => {
            this.colorChange.emit(getValueByType(value, this.childControl.initType));
        }));
        this.subscriptions.push(this.control.valueChanges.subscribe((value) => {
            this.cdr.markForCheck();
            const presets = this.mapColors[value.toHexString()];
            if (presets) {
                this.childControl.setColorPresets(presets);
            }
            this.colorChange.emit(getValueByType(this.childControl.value, this.childControl.initType));
        }));
    }
    ngOnDestroy() {
        this.cdr.detach();
        this.subscriptions.forEach((subscription) => subscription.unsubscribe());
        this.subscriptions.length = 0;
    }
    ngOnChanges(changes) {
        if (this.color && this.control && getValueByType(this.control.value, this.control.initType) !== this.color) {
            this.childControl.setValueFrom(this.color);
        }
    }
}
SwatchesPickerComponent.ɵfac = function SwatchesPickerComponent_Factory(t) { return new (t || SwatchesPickerComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
SwatchesPickerComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: SwatchesPickerComponent, selectors: [["swatches-picker"]], inputs: { color: "color" }, outputs: { colorChange: "colorChange" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 2, vars: 4, consts: [["direction", "down", 3, "columns", "color", "colorPresets", "colorChange"], ["class", "child-list", "direction", "down", 3, "columns", "color", "colorPresets", "colorChange", 4, "ngIf"], ["direction", "down", 1, "child-list", 3, "columns", "color", "colorPresets", "colorChange"]], template: function SwatchesPickerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "color-presets-component", 0);
        ɵngcc0.ɵɵlistener("colorChange", function SwatchesPickerComponent_Template_color_presets_component_colorChange_0_listener($event) { return ctx.control.value = $event; });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(1, SwatchesPickerComponent_color_presets_component_1_Template, 1, 3, "color-presets-component", 1);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("columns", 7)("color", ctx.control.value)("colorPresets", ctx.control.presets);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.childControl.presets.length);
    } }, directives: [ɵngcc1.ColorPresetsComponent, ɵngcc2.NgIf], styles: ["[_nghost-%COMP%], [_nghost-%COMP%]     *{padding:0;margin:0;box-sizing:border-box}", "@charset \"UTF-8\";[_nghost-%COMP%]{display:block;background:#fff;width:224px;border:1px solid rgba(0,0,0,.2);box-shadow:0 0 2px rgba(0,0,0,.3);border-radius:4px;padding:6px}[_nghost-%COMP%]     color-preset, [_nghost-%COMP%]     color-preset-sublist{width:30px;height:30px}[_nghost-%COMP%]     color-preset.selected:after, [_nghost-%COMP%]     color-preset:hover:after{display:block;content:\"\u00A0\";position:absolute;left:3px;top:3px;bottom:3px;right:3px;z-index:10;border:3px solid #fff;box-shadow:0 0 5px 2px rgba(0,0,0,.2)}[_nghost-%COMP%]     .row:first-child color-preset-sublist:first-child, [_nghost-%COMP%]     .row:first-child color-preset:first-child{border-radius:4px 0 0 4px}[_nghost-%COMP%]     .row.last color-preset-sublist.last, [_nghost-%COMP%]     .row.last color-preset.last{border-radius:0 4px 4px 0}[_nghost-%COMP%]     .child-list .row.first color-preset-sublist.first, [_nghost-%COMP%]     .child-list .row.first color-preset.first{border-radius:4px 0 0 0}[_nghost-%COMP%]     .child-list .row.first color-preset-sublist.last, [_nghost-%COMP%]     .child-list .row.first color-preset.last{border-radius:0 4px 0 0}[_nghost-%COMP%]     .child-list .row.last color-preset-sublist.first, [_nghost-%COMP%]     .child-list .row.last color-preset.first{border-radius:0 0 0 4px}[_nghost-%COMP%]     .child-list .row.last color-preset-sublist.last, [_nghost-%COMP%]     .child-list .row.last color-preset.last{border-radius:0 0 4px 0}[_nghost-%COMP%]     .child-list{margin-top:6px;border-top:1px solid #e5e5e5;padding:6px 0 0}[_nghost-%COMP%]     .child-list color-preset.selected:after, [_nghost-%COMP%]     .child-list color-preset:hover:after{content:\"\u2714\";font-size:18px;color:#fff;border:0;left:0;top:0;bottom:0;right:0;line-height:30px;box-shadow:none;text-align:center}"], changeDetection: 0 });
SwatchesPickerComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
SwatchesPickerComponent.propDecorators = {
    color: [{ type: Input }],
    colorChange: [{ type: Output }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(SwatchesPickerComponent, [{
        type: Component,
        args: [{
                selector: `swatches-picker`,
                template: "<color-presets-component \r\n    [columns]=\"7\"\r\n    direction=\"down\"\r\n    [(color)]=\"control.value\" \r\n    [colorPresets]=\"control.presets\"></color-presets-component>\r\n<color-presets-component\r\n    class=\"child-list\"\r\n    *ngIf=\"childControl.presets.length\"\r\n    [columns]=\"7\"\r\n    direction=\"down\"\r\n    [(color)]=\"childControl.value\" \r\n    [colorPresets]=\"childControl.presets\"></color-presets-component>",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [":host,:host ::ng-deep *{padding:0;margin:0;box-sizing:border-box}", "@charset \"UTF-8\";:host{display:block;background:#fff;width:224px;border:1px solid rgba(0,0,0,.2);box-shadow:0 0 2px rgba(0,0,0,.3);border-radius:4px;padding:6px}:host ::ng-deep color-preset,:host ::ng-deep color-preset-sublist{width:30px;height:30px}:host ::ng-deep color-preset.selected:after,:host ::ng-deep color-preset:hover:after{display:block;content:\"\u00A0\";position:absolute;left:3px;top:3px;bottom:3px;right:3px;z-index:10;border:3px solid #fff;box-shadow:0 0 5px 2px rgba(0,0,0,.2)}:host ::ng-deep .row:first-child color-preset-sublist:first-child,:host ::ng-deep .row:first-child color-preset:first-child{border-radius:4px 0 0 4px}:host ::ng-deep .row.last color-preset-sublist.last,:host ::ng-deep .row.last color-preset.last{border-radius:0 4px 4px 0}:host ::ng-deep .child-list .row.first color-preset-sublist.first,:host ::ng-deep .child-list .row.first color-preset.first{border-radius:4px 0 0 0}:host ::ng-deep .child-list .row.first color-preset-sublist.last,:host ::ng-deep .child-list .row.first color-preset.last{border-radius:0 4px 0 0}:host ::ng-deep .child-list .row.last color-preset-sublist.first,:host ::ng-deep .child-list .row.last color-preset.first{border-radius:0 0 0 4px}:host ::ng-deep .child-list .row.last color-preset-sublist.last,:host ::ng-deep .child-list .row.last color-preset.last{border-radius:0 0 4px 0}:host ::ng-deep .child-list{margin-top:6px;border-top:1px solid #e5e5e5;padding:6px 0 0}:host ::ng-deep .child-list color-preset.selected:after,:host ::ng-deep .child-list color-preset:hover:after{content:\"\u2714\";font-size:18px;color:#fff;border:0;left:0;top:0;bottom:0;right:0;line-height:30px;box-shadow:none;text-align:center}"]
            }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }]; }, { colorChange: [{
            type: Output
        }], color: [{
            type: Input
        }] }); })();
