// @iplab/ngx-color-picker/esm2015/lib/components/parts/inputs/hsla-input/hsla-input.component.js
import { Component, Input, EventEmitter, Output, ChangeDetectionStrategy } from '@angular/core';
import { Color } from './../../../../helpers/color.class';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../../../directives/color-picker-input.directive';
import * as ɵngcc2 from '@angular/common';

function HslaComponent_span_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1, "H");
    ɵngcc0.ɵɵelementEnd();
} }
function HslaComponent_span_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1, "S");
    ɵngcc0.ɵɵelementEnd();
} }
function HslaComponent_span_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1, "L");
    ɵngcc0.ɵɵelementEnd();
} }
function HslaComponent_div_9_span_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1, "A");
    ɵngcc0.ɵɵelementEnd();
} }
function HslaComponent_div_9_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 0);
    ɵngcc0.ɵɵelementStart(1, "input", 5);
    ɵngcc0.ɵɵlistener("inputChange", function HslaComponent_div_9_Template_input_inputChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r5 = ɵngcc0.ɵɵnextContext(); return ctx_r5.onInputChange($event, "A"); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(2, HslaComponent_div_9_span_2_Template, 2, 0, "span", 2);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("value", ctx_r3.value == null ? null : ctx_r3.value.getAlpha().toString());
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.labelVisible);
} }
export class HslaComponent {
    constructor() {
        this.hueChange = new EventEmitter(false);
        this.colorChange = new EventEmitter(false);
        this.isAlphaVisible = true;
    }
    set label(value) {
        this.labelVisible = true;
    }
    set alpha(isVisible) {
        this.isAlphaVisible = isVisible;
    }
    get value() {
        return this.color ? this.color.getHsla() : null;
    }
    onInputChange(newValue, color) {
        const value = this.value;
        const hue = color === 'H' ? newValue : value.hue;
        const saturation = color === 'S' ? newValue : value.saturation;
        const lightness = color === 'L' ? newValue : value.lightness;
        const alpha = color === 'A' ? newValue : value.alpha;
        const newColor = new Color().setHsla(hue, saturation, lightness, alpha);
        const hueColor = new Color().setHsva(newColor.getHsva().hue);
        this.hueChange.emit(hueColor);
        this.colorChange.emit(newColor);
    }
}
HslaComponent.ɵfac = function HslaComponent_Factory(t) { return new (t || HslaComponent)(); };
HslaComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: HslaComponent, selectors: [["hsla-input-component"]], inputs: { label: "label", alpha: "alpha", hue: "hue", color: "color" }, outputs: { hueChange: "hueChange", colorChange: "colorChange" }, decls: 10, vars: 7, consts: [[1, "column"], ["type", "text", "pattern", "[0-9]*", "min", "0", "max", "360", 3, "value", "inputChange"], [4, "ngIf"], ["type", "text", "pattern", "[0-9]*", "min", "0", "max", "100", 3, "value", "inputChange"], ["class", "column", 4, "ngIf"], ["type", "text", "pattern", "[0-9]+([\\.,][0-9]{1,2})?", "min", "0", "max", "1", 3, "value", "inputChange"]], template: function HslaComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "input", 1);
        ɵngcc0.ɵɵlistener("inputChange", function HslaComponent_Template_input_inputChange_1_listener($event) { return ctx.onInputChange($event, "H"); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(2, HslaComponent_span_2_Template, 2, 0, "span", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "div", 0);
        ɵngcc0.ɵɵelementStart(4, "input", 3);
        ɵngcc0.ɵɵlistener("inputChange", function HslaComponent_Template_input_inputChange_4_listener($event) { return ctx.onInputChange($event, "S"); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(5, HslaComponent_span_5_Template, 2, 0, "span", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "div", 0);
        ɵngcc0.ɵɵelementStart(7, "input", 3);
        ɵngcc0.ɵɵlistener("inputChange", function HslaComponent_Template_input_inputChange_7_listener($event) { return ctx.onInputChange($event, "L"); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(8, HslaComponent_span_8_Template, 2, 0, "span", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(9, HslaComponent_div_9_Template, 3, 2, "div", 4);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("value", ctx.value == null ? null : ctx.value.getHue().toString());
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.labelVisible);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("value", (ctx.value == null ? null : ctx.value.getSaturation()) + "%");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.labelVisible);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("value", (ctx.value == null ? null : ctx.value.getLightness()) + "%");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.labelVisible);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isAlphaVisible);
    } }, directives: [ɵngcc1.ColorPickerInputDirective, ɵngcc2.NgIf], styles: ["[_nghost-%COMP%], [_nghost-%COMP%]     *{padding:0;margin:0;box-sizing:border-box}", "[_nghost-%COMP%]{display:table;width:100%;text-align:center;color:#b4b4b4;font-size:11px}.column[_ngcontent-%COMP%]{display:table-cell;padding:0 2px}input[_ngcontent-%COMP%]{width:100%;border:1px solid #dadada;color:#272727;text-align:center;font-size:12px;-webkit-appearance:none;border-radius:0;margin:0 0 6px;height:26px;outline:none}", ""], changeDetection: 0 });
HslaComponent.propDecorators = {
    hue: [{ type: Input }],
    hueChange: [{ type: Output }],
    color: [{ type: Input }],
    colorChange: [{ type: Output }],
    label: [{ type: Input }],
    alpha: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(HslaComponent, [{
        type: Component,
        args: [{
                selector: `hsla-input-component`,
                template: "<div class=\"column\">\r\n    <input type=\"text\" pattern=\"[0-9]*\" min=\"0\" max=\"360\" [value]=\"value?.getHue().toString()\" (inputChange)=\"onInputChange($event, 'H')\" />\r\n    <span *ngIf=\"labelVisible\">H</span>\r\n</div>\r\n<div class=\"column\">\r\n    <input type=\"text\" pattern=\"[0-9]*\" min=\"0\" max=\"100\" [value]=\"value?.getSaturation() + '%'\" (inputChange)=\"onInputChange($event, 'S')\" />\r\n    <span *ngIf=\"labelVisible\">S</span>\r\n</div>\r\n<div class=\"column\">\r\n    <input type=\"text\" pattern=\"[0-9]*\" min=\"0\" max=\"100\" [value]=\"value?.getLightness() + '%'\" (inputChange)=\"onInputChange($event, 'L')\" />\r\n    <span *ngIf=\"labelVisible\">L</span>\r\n</div>\r\n<div class=\"column\" *ngIf=\"isAlphaVisible\">\r\n    <input type=\"text\" pattern=\"[0-9]+([\\.,][0-9]{1,2})?\" min=\"0\" max=\"1\" [value]=\"value?.getAlpha().toString()\" (inputChange)=\"onInputChange($event, 'A')\" />\r\n    <span *ngIf=\"labelVisible\">A</span>\r\n</div>",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [":host,:host ::ng-deep *{padding:0;margin:0;box-sizing:border-box}", ":host{display:table;width:100%;text-align:center;color:#b4b4b4;font-size:11px}.column{display:table-cell;padding:0 2px}input{width:100%;border:1px solid #dadada;color:#272727;text-align:center;font-size:12px;-webkit-appearance:none;border-radius:0;margin:0 0 6px;height:26px;outline:none}", ""]
            }]
    }], function () { return []; }, { hueChange: [{
            type: Output
        }], colorChange: [{
            type: Output
        }], label: [{
            type: Input
        }], alpha: [{
            type: Input
        }], hue: [{
            type: Input
        }], color: [{
            type: Input
        }] }); })();
