// @iplab/ngx-color-picker/esm2015/lib/directives/color-picker-input.directive.js
import { Directive, Input, Output, EventEmitter, HostListener } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class ColorPickerInputDirective {
    constructor() {
        this.inputChange = new EventEmitter();
    }
    inputChanges(event) {
        const element = event.target || event.srcElement;
        const value = element.value;
        const numeric = parseFloat(value);
        if (!isNaN(numeric) && numeric >= parseInt(this.min, 10) &&
            numeric <= parseInt(this.max, 10)) {
            this.inputChange.emit(numeric);
        }
    }
}
ColorPickerInputDirective.ɵfac = function ColorPickerInputDirective_Factory(t) { return new (t || ColorPickerInputDirective)(); };
ColorPickerInputDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ColorPickerInputDirective, selectors: [["", "inputChange", ""]], hostBindings: function ColorPickerInputDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("input", function ColorPickerInputDirective_input_HostBindingHandler($event) { return ctx.inputChanges($event); });
    } }, inputs: { min: "min", max: "max" }, outputs: { inputChange: "inputChange" } });
ColorPickerInputDirective.propDecorators = {
    min: [{ type: Input }],
    max: [{ type: Input }],
    inputChange: [{ type: Output }],
    inputChanges: [{ type: HostListener, args: ['input', ['$event'],] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ColorPickerInputDirective, [{
        type: Directive,
        args: [{
                selector: '[inputChange]'
            }]
    }], function () { return []; }, { inputChange: [{
            type: Output
        }], inputChanges: [{
            type: HostListener,
            args: ['input', ['$event']]
        }], min: [{
            type: Input
        }], max: [{
            type: Input
        }] }); })();
