// mat-video/esm5/lib/ui/mat-download-button/mat-download-button.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from "@angular/core";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/material/button';
import * as ɵngcc2 from '@angular/material/icon';
var MatDownloadButtonComponent = /** @class */ (function () {
    function MatDownloadButtonComponent() {
    }
    /** @nocollapse */
    MatDownloadButtonComponent.ctorParameters = function () { return []; };
    MatDownloadButtonComponent.propDecorators = {
        video: [{ type: Input }],
        title: [{ type: Input }]
    };
MatDownloadButtonComponent.ɵfac = function MatDownloadButtonComponent_Factory(t) { return new (t || MatDownloadButtonComponent)(); };
MatDownloadButtonComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: MatDownloadButtonComponent, selectors: [["mat-download-button"]], inputs: { video: "video", title: "title" }, decls: 3, vars: 2, consts: [["mat-icon-button", "", 3, "href", "download"]], template: function MatDownloadButtonComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "a", 0);
        ɵngcc0.ɵɵelementStart(1, "mat-icon");
        ɵngcc0.ɵɵtext(2, "file_download");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("href", ctx.video == null ? null : ctx.video.currentSrc, ɵngcc0.ɵɵsanitizeUrl)("download", ctx.title);
    } }, directives: [ɵngcc1.MatAnchor, ɵngcc2.MatIcon], styles: ["a[_ngcontent-%COMP%]{color:inherit;text-decoration:none}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatDownloadButtonComponent, [{
        type: Component,
        args: [{
                selector: "mat-download-button",
                template: "<a mat-icon-button [href]=\"video?.currentSrc\" [download]=\"title\">\r\n  <mat-icon>file_download</mat-icon>\r\n</a>\r\n",
                styles: ["a{color:inherit;text-decoration:none}"]
            }]
    }], function () { return []; }, { video: [{
            type: Input
        }], title: [{
            type: Input
        }] }); })();
    return MatDownloadButtonComponent;
}());
export { MatDownloadButtonComponent };
if (false) {
    /** @type {?} */
    MatDownloadButtonComponent.prototype.video;
    /** @type {?} */
    MatDownloadButtonComponent.prototype.title;
}
